### PR TITLE
Testscripts

### LINKED ISSUE #285: Flexibility in course test scripts
We can make the test template scripts, e.g. `go.sh` and friends, be more flexibly defined by adding them to the `tests` repo for the course. Then we don't need to put them in `ci/scripts`. We check for `tests/scripts/*.sh` file; we should probably allow only a single file. If this exists in the `tests` repo, we can save the script in the database (for the course in question), and update it each time the `tests` repo is updated (we already catch the push events from github.)

This will allow teachers to hack and debug their own scripts more easily.

### PR DESCRIPTION
**Pull request by 0xf8f8ff**
New actions on push event to the `Tests` repository:
- if `labX/criteria.json` found: saves grading criteria for the assignment into the database
- if `scripts/run.sh` found: saves as `Assignment.Scriptfile` for each course assignment to be used instead of `quickfeed/ci/scripts/_.sh` when running tests
- if `labX/run.sh` found: save and use when running tests for this assignment (instead of the default one)
- if `scripts/Dockerfile` found: save into the database and build the image locally (tagged with the course code like `DAT123`)

Resolves #285, #477.

### CODE REVIEW

File: database/gormdb_submission.go [hunk 1]
```diff
@@ -65,13 +66,20 @@ func (db *GormDB) CreateSubmission(submission *pb.Submission) error {
 	// TODO(meling) temporary transformation of submission data
 	transform(submission)
 
-	if submission.BuildInfo != nil {
-		if err := db.conn.Save(submission.BuildInfo).Error; err != nil {
+	// A submission will have a build info without any ID, using save will create a duplicate build info
+	// in case the submission already exists in the database. We have to update build info explicitly.
+	// There should be a less hacky way to do it.
```
Comment (meling) [inline database/gormdb_submission.go:71]: Not sure how to do it less hacky. But is there a test that exposes this problem? If not, we should add one. I added print statements inside the if-block below, and no prints where found during testing. So suspect there isn't any tests yet...

I think the reason could be that I may have removed the `gorm:foreignKey` tag in `score.proto`:

```proto
    uint64 SubmissionID = 2 [(go.field) = {tags: 'gorm:"foreignKey:ID"'}];
```

I used to have this, but I have been hacking on this stuff, and I might have removed it without any of the tests breaking, and thereby thinking it would be okay...

Note also, in `web/submissions_test.go:L433-434,441-442` we actually set the `ID` and `SubmissionID`. These should not be set when calling `CreateSubmission`, I think... All tests still passes, when I remove these.

File: doc/teacher.md [hunk 2]
```diff
@@ -163,22 +163,30 @@ This is the purpose of the `tests` repository.
 
 The file system layout of the `tests` repository must match that of the `assignments` repository, as shown below.
 The `assignment.yml` files contains the [assignment information](#assignment-information).
-In addition, each folder should also contain test code for the corresponding assignment.
+In addition, each assignment folder should also contain test code for the corresponding assignment.
+The `scripts` folder must contain a `run.sh` script file with commands to be executed when running assignment tests. This script will be used for every course assignment unless there is an assignment-specific `run.sh` script in the assignment folder. This script will only be used when running tests for this specific assignment.
```
Comment (meling) [inline doc/teacher.md:167]: Please use a new line for each sentence.

I would rephrase it some: 
- The `scripts` folder may contain a `run.sh` script with commands ...
(must is too strong I think, since it is optional if all other assignments have a `run.sh` script).
- This script will only be used --> An assignment-specific `run.sh` script will only be used to run tests for the specific assignment.

File: doc/teacher.md [hunk 3]
```diff
@@ -228,7 +235,7 @@ Comments can be left to every criterion checkpoint or to the whole group of grad
 
 It is also possible to mass approve submissions or mass release reviews for an assignment by choosing a minimal score and then pressing `Approve all` or `Release all` correspondingly. Every submission with a score equal or above the set minimal score will be approved or reviews to such submissions will be released.
 
-Grading criteria can be loaded from a file `criteria.json` in a corresponding assignment folder inside the `Tests` repository.
+Grading criteria will be loaded from a `criteria.json` file if it is added to the corresponding assignment folder inside the `Tests` repository.
```
Comment (meling) [inline doc/teacher.md:238]: Prefer to use lower case repo names: `tests`.

**Review by meling — CHANGES_REQUESTED** [review]

File: ag/ag.proto [hunk 4]
```diff
@@ -76,11 +76,12 @@ message Course {
     uint64 organizationID = 8;
     string organizationPath = 9; // The organization's SCM name, e.g. uis-dat520-2020.
     uint32 slipDays = 10;
-    Enrollment.UserStatus enrolled = 11 [(go.field) = {tags: 'sql:\"-\"'}];
+    string dockerfile = 11;
+    Enrollment.UserStatus enrolled = 12 [(go.field) = {tags: 'sql:\"-\"'}];
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ag/ag.proto:80]: Done.

File: assignments/assignments.go [hunk 5]
```diff
@@ -21,29 +25,35 @@ func UpdateFromTestsRepo(logger *zap.SugaredLogger, db database.Database, repo *
...
 		return
 	}
 	for _, assignment := range assignments {
-		logger.Debugf("Found assignment in '%s' repository: %v", pb.TestsRepo, assignment)
+		updateGradingCriteria(logger, db, assignment)
+	}
+	if dockerfile != "" && dockerfile != course.Dockerfile {
+		course.Dockerfile = dockerfile
+		if err := db.UpdateCourse(course); err != nil {
+			logger.Debugf("Failed to update Dockerfile for course %s: %s", course.Code, err)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments.go:39]: Fixed.

File: assignments/assignments.go [hunk 6]
```diff
@@ -21,29 +25,35 @@ func UpdateFromTestsRepo(logger *zap.SugaredLogger, db database.Database, repo *
 		logger.Errorf("Failed to create SCM Client: %v", err)
 		return
 	}
-	assignments, err := FetchAssignments(context.Background(), s, course)
+	assignments, dockerfile, err := FetchAssignments(context.Background(), s, course)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments.go:28]: Done.

File: kit/score/score.pb.go [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.26.0
-// 	protoc        v3.15.6
+// 	protoc-gen-go v1.27.1
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply kit/score/score.pb.go:3]: Right, turned out I had two conflicting protoc versions. Fixed now.

File: web/assignments.go [hunk 8]
```diff
@@ -38,10 +35,16 @@ func (s *AutograderService) updateAssignments(ctx context.Context, sc scm.SCM, c
 	if err != nil {
 		return err
 	}
-	assignments, err := assignments.FetchAssignments(ctx, sc, course)
+	assignments, dockerfile, err := assignments.FetchAssignments(ctx, sc, course)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/assignments.go:38]: In progress.

File: assignments/assignments_parser.go [hunk 9]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
...
+				courseDockerfile = buffer.String()
 			}
 		}
 		return nil
 	})
 	if err != nil {
-		return nil, err
+		return nil, "", err
+	}
+	if defaultScript != "" {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:139]: Added.

File: assignments/assignments_parser.go [hunk 10]
```diff
@@ -37,17 +42,20 @@ type assignmentData struct {
...
 
 	var assignments []*pb.Assignment
+	var defaultScript string
+	var courseDockerfile string
 	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
 		if !info.IsDir() {
 			filename := filepath.Base(path)
-			if filename == target || filename == targetYaml {
+			switch filename {
+			case target, targetYaml:
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:58]: Done.

File: assignments/assignments_parser.go [hunk 11]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
...
+				}
+				if currentAssignmentName == scriptFolder {
+					defaultScript = string(content)
+				} else {
+					assignment := findAssignmentByName(assignments, currentAssignmentName)
+					if assignment == nil {
+						return fmt.Errorf("Found scriptfile in assignment %s folder, could not find thr assignment\n", currentAssignmentName)
+					}
+					assignment.ScriptFile = string(content)
+				}
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:121]: Fixed.

File: assignments/assignments_parser.go [hunk 12]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
 				}
 
 				assignments = append(assignments, assignment)
+
+			case criteriaFile:
+				assignmentName := filepath.Base(filepath.Dir(path))
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:91]: Done.

File: assignments/assignments_parser.go [hunk 13]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
...
+					return fmt.Errorf("Could not to read %q file: %w", filename, err)
+				}
+				var benchmarks []*pb.GradingBenchmark
+				if err := json.Unmarshal(criteria, &benchmarks); err != nil {
+					return fmt.Errorf("Error unmarshalling criteria.json: %s", err)
+				}
+				assignment := findAssignmentByName(assignments, assignmentName)
+				if assignment == nil {
+					return fmt.Errorf("Found benchmarks for assignment %s, could not find the assignment\n", assignmentName)
+				} else {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:103]: Fixed.

File: assignments/assignments_parser.go [hunk 14]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
...
+				if err != nil {
+					return fmt.Errorf("Could not to read %q file: %w", filename, err)
+				}
+				var benchmarks []*pb.GradingBenchmark
+				if err := json.Unmarshal(criteria, &benchmarks); err != nil {
+					return fmt.Errorf("Error unmarshalling criteria.json: %s", err)
+				}
+				assignment := findAssignmentByName(assignments, assignmentName)
+				if assignment == nil {
+					return fmt.Errorf("Found benchmarks for assignment %s, could not find the assignment\n", assignmentName)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:102]: Fixed for all cases.

File: assignments/assignments_parser.go [hunk 15]
```diff
@@ -82,14 +86,64 @@ func parseAssignments(dir string, courseID uint64) ([]*pb.Assignment, error) {
...
+					defaultScript = string(content)
+				} else {
+					assignment := findAssignmentByName(assignments, currentAssignmentName)
+					if assignment == nil {
+						return fmt.Errorf("Found scriptfile in assignment %s folder, could not find thr assignment\n", currentAssignmentName)
+					}
+					assignment.ScriptFile = string(content)
+				}
+			case dockerfile:
+				t, err := template.ParseFiles(path)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser.go:123]: Right, we don't need template here, replaced by just reading and saving file contents.

File: assignments/assignments_parser_test.go [hunk 16]
```diff
@@ -73,40 +101,87 @@ func TestParse(t *testing.T) {
...
 		t.Errorf("len(assignments) = %d, want %d", len(assignments), 2)
 	}
+	if dockerfile != df {
+		t.Errorf("Incorrect dockerfile\n Want: %s\n Got: %s\n", df, dockerfile)
+	}
 	if diff := cmp.Diff(assignments[0], wantAssignment1, cmpopts.IgnoreUnexported(pb.Assignment{})); diff != "" {
 		t.Errorf("parseAssignments() mismatch (-want +got):\n%s", diff)
 	}
-	if diff := cmp.Diff(assignments[1], wantAssignment2, cmpopts.IgnoreUnexported(pb.Assignment{})); diff != "" {
+	if diff := cmp.Diff(assignments[1], wantAssignment2, cmpopts.IgnoreUnexported(pb.Assignment{}), protocmp.Transform()); diff != "" {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments_parser_test.go:179]: Removed.

File: ci/docker.go [hunk 17]
```diff
@@ -170,30 +169,26 @@ func (d *Docker) pullImage(ctx context.Context, image string) error {
...
-	d.logger.Infof("Building image: '%s' from %s", image, dockerfile)
-
-	contents, err := ioutil.ReadFile(dockerfile)
-	if err != nil {
-		return err
+func (d *Docker) buildImage(ctx context.Context, dockerfile string, image string) error {
+	if dockerfile == "" || image == "" {
+		return fmt.Errorf("Failed to build image %s: missing Dockerfile in Tests/scripts/ or image name in run.sh", image)
 	}
+	dockerbytes := []byte(dockerfile)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ci/docker.go:176]: Changed.

File: ci/docker_test.go [hunk 18]
```diff
@@ -37,8 +37,9 @@ func TestDocker(t *testing.T) {
 	}
 
 	const (
-		script  = `echo -n "hello world"`
-		wantOut = "hello world"
+		script     = `echo -n "hello world"`
+		wantOut    = "hello world"
+		dockerfile = "FROM golang:1.17beta1-alpine\n WORKDIR /quickfeed"
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ci/docker_test.go:42]: Replaced.

File: ci/parse_script.go [hunk 19]
```diff
@@ -40,9 +34,8 @@ func newAssignmentInfo(course *pb.Course, assignment *pb.Assignment, cloneURL, t
 // parseScriptTemplate returns a job describing the docker image to use and
 // the commands of the job. The job is extracted from a script template file
 // provided as input along with assignment metadata for the template.
-func parseScriptTemplate(scriptPath string, info *AssignmentInfo) (*Job, error) {
-	tmplFile := filepath.Join(scriptPath, info.Script)
-	t, err := template.ParseFiles(tmplFile)
+func parseScriptTemplate(info *AssignmentInfo) (*Job, error) {
+	t, err := template.New("scriptfile").Parse(info.Script)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply ci/parse_script.go:38]: Added.

File: database/gormdb_submission.go [hunk 20]
```diff
@@ -65,13 +66,20 @@ func (db *GormDB) CreateSubmission(submission *pb.Submission) error {
...
+	if submission.BuildInfo != nil && submission.ID != 0 {
+		if err := db.conn.Where(&score.BuildInfo{SubmissionID: submission.ID}).Updates(&submission.BuildInfo).Error; err != nil {
 			return err
 		}
+		var buildinfo score.BuildInfo
+		if err := db.conn.Where(&score.BuildInfo{SubmissionID: submission.ID}).Last(&buildinfo).Error; err != nil {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply database/gormdb_submission.go:77]: Done.

File: database/gormdb_submission.go [hunk 21]
```diff
@@ -65,13 +66,20 @@ func (db *GormDB) CreateSubmission(submission *pb.Submission) error {
...
-	if submission.BuildInfo != nil {
-		if err := db.conn.Save(submission.BuildInfo).Error; err != nil {
+	// A submission will have a build info without any ID, using save will create a duplicate build info
+	// in case the submission already exists in the database. We have to update build info explicitly.
+	// There should be a less hacky way to do it.
+	if submission.BuildInfo != nil && submission.ID != 0 {
+		if err := db.conn.Where(&score.BuildInfo{SubmissionID: submission.ID}).Updates(&submission.BuildInfo).Error; err != nil {
 			return err
 		}
+		var buildinfo score.BuildInfo
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply database/gormdb_submission.go:76]: Done.

File: database/gormdb_submission.go (hunk 1 above)
Comment (0xf8f8ff) [reply database/gormdb_submission.go:71]: It is something I've noticed about the gorm's Save() behavior: if we save a struct that has another struct (or a slice of these) as its field (like `Submission.BuildInfo`), Save will properly create records for the nested struct(s) if there are no primary IDs in those, but just ignores them (neither creating nor updating) when there are primary IDs. Doesn't look like it has something with foreign keys to do.

Added a new test for this behavior.

File: doc/teacher.md (hunk 2 above)
Comment (0xf8f8ff) [reply doc/teacher.md:167]: Updated.

File: doc/teacher.md (hunk 3 above)
Comment (0xf8f8ff) [reply doc/teacher.md:238]: Fixed.

File: web/assignments.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply web/assignments.go]: Done: `UpdateFromtestRepo` could actually be used in both cases, there was no need for a separate function as `updateAssignments` since we do not really need to pass a repository to the former anymore.

File: assignments/assignments.go [hunk 22]
```diff
@@ -21,29 +25,35 @@ func UpdateFromTestsRepo(logger *zap.SugaredLogger, db database.Database, repo *
...
 	if err != nil {
 		logger.Errorf("Failed to fetch assignments from '%s' repository: %v", pb.TestsRepo, err)
 		return
 	}
 	for _, assignment := range assignments {
-		logger.Debugf("Found assignment in '%s' repository: %v", pb.TestsRepo, assignment)
+		updateGradingCriteria(logger, db, assignment)
+	}
+	if dockerfile != "" && dockerfile != course.Dockerfile {
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments.go:36]: Not sure if it is worth it to pass the `db` but we could just return a course and error here with `Course.Assignments` and `Course.Dockerfile`.

File: assignments/assignments.go [hunk 23]
```diff
@@ -93,9 +91,82 @@ func FetchAssignments(c context.Context, sc scm.SCM, course *pb.Course) ([]*pb.A
...
+	// if a Dockerfile added/updated, build docker image locally
+	// tag the image with the course code
+	if dockerfile != "" && dockerfile != course.Dockerfile {
+		job.Commands = []string{
+			"cd " + cloneDir + "/tests/scripts",
+			fmt.Sprintf("docker build -t %s .", course.Code),
+		}
+
+		if out, err := runner.Run(context.Background(), job); err != nil {
+			log.Printf("Failed to build image from dockerfile for %s (%s): %s", course.Code, out, err)
```
[the remark this answers is not in the document]
Comment (0xf8f8ff) [reply assignments/assignments.go:112]: Done.

**Review by meling — APPROVED** [review]
